mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java

```diff
@@ -145,13 +145,24 @@ private static UidCriterion findConjugatedUidCriterion(List<Criterion> crits) {
 
     @Override
     public List<MessageId> search(MailboxSession session, final Collection<MailboxId> mailboxIds, SearchQuery searchQuery, long limit) throws MailboxException {
+        return getAsMessageIds(searchAllMessageInMailboxes(session, mailboxIds, searchQuery), limit);

```

Comment:
We IMO should hace a collector building the limited result list AND counting matches at the same time. This way we would gain performance and consistency.